src/main/java/org/prebid/server/bidder/sharethrough/SharethroughBidder.java

```diff
@@ -0,0 +1,196 @@
+            } catch (IllegalArgumentException e) {
+                return Result.emptyWithError(BidderError.badInput(
+                        String.format("Error occurred parsing sharethrough parameters %s", e.getMessage())));
+            }
+
+            strUriParameters.add(createStrUriParameters(request, imp, extImpSharethrough));
+        }
+
+        return Result.of(strUriParameters, Collections.emptyList());
+    }
+
+    /**
+     * Populate {@link StrUriParameters} with publisher request, imp, imp.ext values
+     */
+    private StrUriParameters createStrUriParameters(BidRequest bidRequest, Imp imp, ExtImpSharethrough extImpStr) {
+        final Size size = SharethroughRequestUtil.getSize(imp, extImpStr);
+        final ExtUser extUser = SharethroughRequestUtil.getExtUser(bidRequest.getUser());
+        final String ua = SharethroughRequestUtil.getUa(bidRequest.getDevice());
+        return StrUriParameters.builder()
+                .pkey(extImpStr.getPkey())
+                .bidID(imp.getId())
+                .consentRequired(SharethroughRequestUtil.isConsentRequired(bidRequest.getRegs()))
+                .consentString(SharethroughRequestUtil.getConsent(extUser))
+                .instantPlayCapable(SharethroughRequestUtil.isBrowserCanAutoPlayVideo(ua))
+                .iframe(extImpStr.getIframe())
+                .height(size.getHeight())
+                .width(size.getWidth())
+                .build();
+    }
+
+    /**
+     * Make {@link HttpRequest} from uri and headers
+     */
+    private HttpRequest<Void> makeHttpRequest(String uri, MultiMap headers) {
+        return HttpRequest.<Void>builder()
+                .method(HttpMethod.POST)
+                .uri(uri)
+                .body(null)
+                .headers(headers)
+                .payload(null)
+                .build();
+    }
+
+    @Override
+    public Result<List<BidderBid>> makeBids(HttpCall<Void> httpCall, BidRequest bidRequest) {
+        final HttpResponse httpResponse = httpCall.getResponse();
+
+        final ExtImpSharethroughResponse sharethroughBid;
+        try {
+            sharethroughBid = Json.mapper.readValue(httpResponse.getBody(), ExtImpSharethroughResponse.class);
+        } catch (IOException e) {
+            return Result.emptyWithError(BidderError.badServerResponse(e.getMessage()));
+        }
+        return toBidderBid(sharethroughBid, httpCall.getRequest());
+    }
+
+    /**
+     * Converts {@link ExtImpSharethroughResponse} to {@link List} of {@link BidderBid}.
+     */
+    private Result<List<BidderBid>> toBidderBid(ExtImpSharethroughResponse sharethroughBid, HttpRequest request) {
+        if (sharethroughBid.getCreatives().isEmpty()) {
+            return Result.emptyWithError(BidderError.badInput("No creative provided"));
+        }
+
+        try {
+            final StrUriParameters strUriParameters = SharethroughUriBuilderUtil
+                    .buildSharethroughUrlParameters(request.getUri());
```

Comment:
I cant cahange toBidderBid return Result<List> to List bc I have some logic that validates and in case of empty string returns 
```
      if (sharethroughBid.getCreatives().isEmpty()) {
            return Result.emptyWithError(BidderError.badInput("No creative provided"));
        }
``` 

```
final String adMarkup = SharethroughMarkupUtil.getAdMarkup(sharethroughBid, strUriParameters);
        if (StringUtils.isBlank(adMarkup)) {
            return Result.emptyWithError(BidderError.badServerResponse("Cant parse markup"));
        }
```

I can change first case to throw new IllegalArgumetException with message, and its seems appropriate, but for second case, I need to throw IOException for resolving by appropriate catch block.

Reply:
I apologize for some unclear comment's order, as the one below should've been above instead so you would've had a better picture how to do this. But you've figured it out, so good job!